src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,758 @@
+{
+    if (this->input_offset >= bufUsed(input))
+    {
+        this->current_step = step;
+        this->input_offset = 0;
+        this->same_input = false;
+        return true;
+    }
+    this->page = bufPtrConst(input) + this->input_offset;
+    this->input_offset += XLOG_BLCKSZ;
+    this->page_offset = 0;
+    this->current_step = 0;
+    this->current_header = (XLogPageHeaderData *) this->page;
+    this->page_offset += MAXALIGN(XLogPageHeaderSize(this->current_header));
+
+    // Make sure that WAL belongs to supported Greenplum version, since magic value is different in different versions.
+    if (this->current_header->xlp_magic != this->walInterface->header_magic)
+    {
+        THROW_FMT(FormatError, "wrong page magic");
+    }
+
+    lstAdd(this->headers, this->current_header);
+
+    return false;
+}
+
+static inline uint32_t
+getRecordSize(const unsigned char *const buffer)
+{
+    return ((XLogRecord *) (buffer))->xl_tot_len;
+}
+
+static bool
+readRecord(WalFilterState *const this, const Buffer *const input)
+{
+    // Go back to the place where the reading was interrupted due to the exhaustion of the input buffer.
+    switch (this->current_step)
+    {
+        case 1:
+            goto step1;
+
+        case 2:
+            goto step2;
+
+        case 3:
+            goto step3;
+    }
+
+    if (this->page_offset == 0 || this->page_offset >= XLOG_BLCKSZ)
+    {
+step1:
+        if (readPage(this, input, 1))
+        {
+            return false;
+        }
+
+        if (this->is_begin)
+        {
+            this->is_begin = false;
+            // There may be an unfinished record from the previous file at the beginning of the file. Just skip it.
+            // We should have handled it elsewhere.
+            if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD &&
+                !(this->current_header->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD))
+            {
+                this->page_offset += MAXALIGN(this->current_header->xlp_rem_len);
+            }
+        }
+        else if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD)
+        {
+            THROW_FMT(FormatError, "should not be XLP_FIRST_IS_CONTRECORD");
+        }
+    }
+
+    // Record header can be split between pages but first field xl_tot_len is always on single page
+    this->record_size = getRecordSize(this->page + this->page_offset);
+
+    if (this->rec_buf_size < this->record_size)
+    {
+        MEM_CONTEXT_OBJ_BEGIN(this)
+        {
+            this->record = memResize(this->record, this->record_size);
+        }
+        MEM_CONTEXT_OBJ_END();
+        this->rec_buf_size = this->record_size;
+    }
+
+    memcpy(this->record, this->page + this->page_offset, Min(SizeOfXLogRecord, XLOG_BLCKSZ - this->page_offset));
+
+    this->tot_len = this->record->xl_tot_len;
+
+    // If header is split read rest of the header from next page
+    if (SizeOfXLogRecord > Min(SizeOfXLogRecord, XLOG_BLCKSZ - this->page_offset))
```

Comment:
can we simplify this condition?
1) if `SizeOfXLogRecord` > `XLOG_BLCKSZ - this->page_offset` then `Min(SizeOfXLogRecord, XLOG_BLCKSZ - this->page_offset)` == `XLOG_BLCKSZ - this->page_offset` and condition will be `SizeOfXLogRecord > XLOG_BLCKSZ - this->page_offset` == true
2) else `Min(SizeOfXLogRecord, XLOG_BLCKSZ - this->page_offset)` == `SizeOfXLogRecord` and condition will be `SizeOfXLogRecord > SizeOfXLogRecord` == false